clash-lib/src/Clash/Normalize/Transformations/Inline.hs

```diff
@@ -379,6 +382,45 @@ inlineCleanup (TransformContext is0 _) (Letrec binds body) = do
 inlineCleanup _ e = return e
 {-# SCC inlineCleanup #-}
 
+-- | Takes a binding and collapses its term if it is a noop
+inlineIdentity :: HasCallStack => NormRewrite
+inlineIdentity _ (Letrec binds body) = do
+  (untouched,collapsed) <- partitionEithers <$> mapM identityAnalysis binds
+  if null collapsed
+    then return (Letrec binds body)
+    else changed (Letrec (untouched ++ collapsed) body)
+inlineIdentity _ e = return e
+
+identityAnalysis :: HasCallStack => (Id,Term) -> NormalizeSession (Either (Id,Term) (Id,Term))
+identityAnalysis old@(iD,collectArgs -> (term,partitionEithers -> (termArgs,typeArgs))) =
+  runIdentityAnalysis $ do
+    Prim primInfo <- return term
+    identity      <- getIdentity primInfo
+    let resTy = polyResTyView $ primType primInfo
+    return (iD,primUCo `App` identity `mkTyApps` (resTy:typeArgs))
+  where
+    runIdentityAnalysis =
+      fmap (Maybe.maybe (Left old) Right) . runMaybeT
+
+    getTermArg i = do
+      Monad.guard $ i <= length termArgs - 1
+      return $ termArgs List.!! i
+
+    -- guards that a given term is a noop
+    noopGuard :: Term -> MaybeT NormalizeSession ()
+    noopGuard (Var i) = do
+        binding <- MaybeT $ lookupVarEnv i <$> Lens.use bindings
+        noopGuard $ bindingTerm binding
+    noopGuard (Prim PrimInfo{primWorkInfo=WorkIdentity _ []}) = return ()
+    noopGuard _ = fail "is not a noop"
```

Comment:
This is the same as `MaybeT (return Nothing)`, but given how infrequently we use `fail` someone reading might not immediately realise that and expect it to behave the same as `error`. I think it would be clearer to either use `MaybeT (return Nothing)` or something like `empty` (from `Alternative`)

Reply:
mhm, I did realize that, however I think that `empty` does not clearly communicate that the computation gets short-circuited.
`MaybeT (return Nothing)` on the other hand does that clearly, but I think that is 'clunky'.
I think it's best to use the actual `Monad.guard` here.